File: 300/367 - Valid Perfect Square.c
bool isPerfectSquare(int num) {
    long right = num / 2, left, i;
    
    while (right * right > num)
        right /= 2;
    
    left = right;
    right = right * 2 + 1;
    for (i = left; i <= right; ++i) {
        if (i * i == num)
            return 1;
    }
    
    return 0;
}

File: csharp/367.valid-perfect-square.cs
/*
 * @lc app=leetcode id=367 lang=csharp
 *
 * [367] Valid Perfect Square
 */

using System.Collections.Generic;

public class Solution {

    private static readonly ISet<int> set;

    static Solution() {
        set = new HashSet<int>();
        long val = 1, i = 1;
        do {
            set.Add((int) val);
            ++i;
            val = i * i;
        } while (val < int.MaxValue);
    }

    public bool IsPerfectSquare(int num) {
        return set.Contains(num);
    }
}

